src/main/java/com/minecolonies/core/event/PermissionEventHandler.java

```diff
@@ -0,0 +1,667 @@
+        {
+            return;
+        }
+
+        validatePrecondition(level, event.getPos(), (colony) -> {
+            final Action action = event.getPlacedBlock().getBlock() instanceof AbstractBlockHut ? Action.PLACE_HUTS : Action.PLACE_BLOCKS;
+            if (!colony.getPermissions().hasPermission(player, action))
+            {
+                cancelEvent(event, event.getEntity(), colony, action, event.getPos());
+            }
+        });
+    }
+
+    /**
+     * Cancel an event and record the denial details in the colony's town hall.
+     *
+     * @param event  the event to cancel
+     * @param entity the player whose action was denied
+     * @param colony the colony where the event took place
+     * @param action the action which was denied
+     * @param pos    the location of the action which was denied
+     */
+    private static void cancelEvent(final Event event, @Nullable final Entity entity, final IColony colony, final Action action, final BlockPos pos)
+    {
+        event.setResult(Result.DENY);
+        if (event.isCancelable())
+        {
+            event.setCanceled(true);
+
+            final UUID uuid = entity != null ? entity.getUUID() : null;
+            final String name = entity != null ? entity.getName().getString() : "-";
+            if (colony.hasTownHall())
+            {
+                colony.getBuildingManager().getTownHall().addPermissionEvent(new PermissionEvent(uuid, name, action, pos));
+            }
+
+            if (entity instanceof FakePlayer)
+            {
+                return;
+            }
+
+            if (entity instanceof Player player)
+            {
+                MessageUtils.format(PERMISSION_DENIED).sendTo(player);
+
+                final long worldTime = entity.level.getGameTime();
+                if (!lastPlayerNotificationTick.containsKey(uuid) || lastPlayerNotificationTick.get(uuid) + (TICKS_SECOND * 10) < worldTime)
+                {
+                    lastPlayerNotificationTick.put(uuid, worldTime);
+                    playerAttempts.put(uuid, 0);
+                }
+                else
+                {
+                    if (playerAttempts.merge(uuid, 1, Integer::sum) > 10)
+                    {
+                        playerAttempts.removeInt(uuid);
+                        player.addEffect(new MobEffectInstance(MobEffects.LEVITATION, TICKS_SECOND * 10));
+                    }
+                }
+            }
+        }
+    }
+
+    /**
+     * BlockEvent.BreakEvent handler.
+     *
+     * @param event BlockEvent.BreakEvent
+     */
+    @SubscribeEvent(priority = EventPriority.HIGH)
+    public static void on(final BreakEvent event)
+    {
+        final LevelAccessor world = event.getLevel();
+        if (!(world instanceof Level level) || level.isClientSide)
+        {
+            return;
+        }
+
+        validatePrecondition(level, event.getPos(), (colony) -> {
```

Comment:
I'm not a huge fan of this lambda wrapper. Can we do just Colony colony = validatePreconditionAndFetchColony?

Reply:
yeah, returns null colony is the condition is not met, so if null return